platform-sdk/docs/components/componentFramework.md

```diff
@@ -0,0 +1,200 @@
+barTaskScheduler.getOutputWire().solderTo(bazInputWire);
+```
+
+![Step 5](step5.drawio.svg)
+
+## Features
+
+### Wiring Model
+
+The wiring model tracks all task schedulers and wires that connect them and is capable of doing useful analysis of the
+resulting graph. Most notably, it can produce mermaid style diagrams that show the flow of data through wires and
+components, detect cyclic backpressure that could result in a deadlock, and verify proper wiring of various task
+schedulers according to their restrictions (for example, a concurrent task scheduler is not permitted to be wired to a
+direct scheduler).
+
+### Wire Add Ons
+
+Wire add ons (or transformers) are simple units of logic that transform the data type on a wire to a different type. For
+example, a `WireListSplitter` takes data from a wire with a `List` type and streams the contents of the list, one by
+one, onto another wire. Wire transformers operate on the thread defined in the task scheduler and do not come with their
+own concurrency logic. Some types of transformers are described below:
+
+**Wire Filter:** A predicate that operates on the data type of the output wire. Data is only passed to the filter output wire if the
+predicate passes.
+
+**Wire Transformer:** A function that operates on the data type of the output wire. Data produced by the wire transformer and forwarded on its
+output wire may be a different type that the input.
+
+**Wire List Splitter:** A function that operates on a list of data and streams the individual items on the output wire.
+
+### Backpressure
+
+Backpressure can be configured for some task schedulers. If available and configured, threads are blocked from
+adding data to a component that has reached the limit of unprocessed tasks. This feature prevents too many tasks from
+building up in a given component and signals the components feeding into it to back off so that it can catch up. In
+order to prevent circular data flows that could result in a deadlock of backpressure, task schedulers provide a way to
+bypass the backpressure mechanism. The method `inject()` ignores the number of unprocessed tasks and the maximum
+configured value in order to prevent such deadlocks. **If backpressure is desired, the primary path of data flow must
+always pass data using the backpressure methods. Secondary data flows must use `inject()` to avoid deadlock**. Circular
+data flows that use backpressure are identified by the wiring model, which logs a warning if cyclic backpressure is
+detected.
+
+### Secondary Output Wires
+
+Each task schedulers comes with a built in output wire. If additional output wires are necessary, they can be created
+using the task scheduler, but they must be passed to the business logic in the constructor so that they can be invoked
+directly by the business logic. It is a violation of the framework principles to invoke a secondary output wire from
+anywhere outside the business logic.
```

Comment:
This paragraph here says that it would be a violation to invoke output wires from outside of the business logic. In other words, it basically says that "business logic invokes some output wires."

But somewhere above the document states:
```
Each unit of business logic does not know where it's input comes from or who consumes its output. The threading is handled by its task scheduler, and the connection of inputs and outputs is handled externally. Changes to concurrency and wiring have no impact on the business logic.
```

which basically reads as "business logic has no idea about the specifics of its input and output wires." I.e. the business logic doesn't invoke its output wires directly.

These statements seem to contradict each other.

I suggest to remove mentions of the business logic from this paragraph here because it seems to be talking about the capabilities of the task schedulers only.

Reply:
> Each unit of business logic does not know ... who consumes its output

In this scenario, the business logic still doesn't know who is consuming the data, only that the data needs to be passed to a particular lambda function.

In general, it's acceptable for the business logic to know that it is sending data out over a wire. But we should avoid scenarios where the business logic must be aware of who is listening for that data on the other end of the wire.

In order to clarify this statement, how about if we rewrote this sentence

> It is a violation of the framework principles to invoke a secondary output wire from
anywhere outside the business logic.

into the following:

```
A component is the sole owner of its output wires. Only the business logic of a component should be allowed to push data into its output wires. It is a violation of the framework principles to share ownership of a secondary output wire by allowing anything external to the component to push data onto the output wire.
```

Reply:
There is a distinction between the primary output wire that comes built into the task scheduler and that the task scheduler sends data on, and secondary output wires which must be invoked directly by the business logic and the task schedulers know nothing about. But in either case, the statement quoted says that the business logic does not know WHO/WHAT provides the input or consumes the output. It is only a violation to invoke SECONDARY output wires from outside of the business logic.

I will try expanding this paragraph to make the distinction more clear.